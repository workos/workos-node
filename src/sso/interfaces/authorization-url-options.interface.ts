export interface AuthorizationURLOptions {
  clientId: string;
  connection?: string;
  organization?: string;

  /**
   * @deprecated Please use `organization` instead.
   */
  domain?: string;
  domainHint?: string;
  loginHint?: string;
  provider?: string;
  redirectUri: string;
  state?: string;
}
