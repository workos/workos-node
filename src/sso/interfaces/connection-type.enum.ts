export enum ConnectionType {
  ADFSSAML = 'ADFSSAML',
  AdpOidc = 'AdpOidc',
  Auth0SAML = 'Auth0SAML',
  AzureSAML = 'AzureSAML',
  CasSAML = 'CasSAML',
  ClassLinkSAML = 'ClassLinkSAML',
  CloudflareSAML = 'CloudflareSAML',
  CyberArkSAML = 'CyberArkSAML',
  DuoSAML = 'DuoSAML',
  GenericOIDC = 'GenericOIDC',
  GenericSAML = 'GenericSAML',
  GoogleOAuth = 'GoogleOAuth',
  GoogleSAML = 'GoogleSAML',
  JumpCloudSAML = 'JumpCloudSAML',
  KeycloakSAML = 'KeycloakSAML',
  LastPassSAML = 'LastPassSAML',
  LoginGovOidc = 'LoginGovOidc',
  MagicLink = 'MagicLink',
  MicrosoftOAuth = 'MicrosoftOAuth',
  MiniOrangeSAML = 'MiniOrangeSAML',
  NetIqSAML = 'NetIqSAML',
  OktaSAML = 'OktaSAML',
  OneLoginSAML = 'OneLoginSAML',
  OracleSAML = 'OracleSAML',
  PingFederateSAML = 'PingFederateSAML',
  PingOneSAML = 'PingOneSAML',
  RipplingSAML = 'RipplingSAML',
  SalesforceSAML = 'SalesforceSAML',
  ShibbolethGenericSAML = 'ShibbolethGenericSAML',
  ShibbolethSAML = 'ShibbolethSAML',
  SimpleSamlPhpSAML = 'SimpleSamlPhpSAML',
  VMwareSAML = 'VMwareSAML',
}
