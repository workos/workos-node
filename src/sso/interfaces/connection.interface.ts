import { ConnectionType } from './connection-type.enum';

export interface ConnectionDomain {
  object: 'connection_domain';
  id: string;
  domain: string;
}

export interface Connection {
  object: 'connection';
  id: string;
  organizationId?: string;
  name: string;
  /**
   * @deprecated The connectionType parameter has been deprecated. Please use type.
   */
  connectionType: ConnectionType;
  state: 'draft' | 'active' | 'inactive' | 'validating';
  domains: ConnectionDomain[];
  type: ConnectionType;
  createdAt: string;
  updatedAt: string;
}

export interface ConnectionResponse {
  object: 'connection';
  id: string;
  organization_id?: string;
  name: string;
  connection_type: ConnectionType;
  state: 'draft' | 'active' | 'inactive' | 'validating';
  domains: ConnectionDomain[];
  created_at: string;
  updated_at: string;
}
