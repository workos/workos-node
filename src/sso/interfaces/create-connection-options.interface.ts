export interface CreateConnectionOptions {
  source: string;
}
