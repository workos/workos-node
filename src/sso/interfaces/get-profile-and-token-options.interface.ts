export interface GetProfileAndTokenOptions {
  clientId: string;
  code: string;
}
