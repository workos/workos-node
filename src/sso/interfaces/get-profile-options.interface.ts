export interface GetProfileOptions {
  accessToken: string;
}
