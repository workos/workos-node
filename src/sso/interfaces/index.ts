export * from './authorization-url-options.interface';
export * from './connection-type.enum';
export * from './connection.interface';
export * from './get-profile-and-token-options.interface';
export * from './profile-and-token.interface';
export * from './profile.interface';
