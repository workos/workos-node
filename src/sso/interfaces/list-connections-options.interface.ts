import { PaginationOptions } from '../../common/interfaces/pagination-options.interface';
import { ConnectionType } from './connection-type.enum';

export interface ListConnectionsOptions extends PaginationOptions {
  connectionType?: ConnectionType;
  domain?: string;
  organizationId?: string;
}

export interface SerializedListConnectionsOptions extends PaginationOptions {
  connection_type?: ConnectionType;
  domain?: string;
  organization_id?: string;
}
