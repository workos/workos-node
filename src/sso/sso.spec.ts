import fetch from 'jest-fetch-mock';
import {
  fetchOnce,
  fetchURL,
  fetchHeaders,
  fetchBody,
  fetchSearchParams,
} from '../common/utils/test-utils';

import { WorkOS } from '../workos';
import { ConnectionResponse, ConnectionType } from './interfaces';
import { ListResponse } from '../common/interfaces';

describe('SSO', () => {
  beforeEach(() => fetch.resetMocks());

  const connectionResponse: ConnectionResponse = {
    object: 'connection',
    id: 'conn_123',
    organization_id: 'org_123',
    name: 'Connection',
    connection_type: ConnectionType.OktaSAML,
    state: 'active',
    domains: [],
    created_at: '2023-07-17T20:07:20.055Z',
    updated_at: '2023-07-17T20:07:20.055Z',
  };

  describe('SSO', () => {
    describe('with options', () => {
      it('requests Connections with query parameters', async () => {
        const workos = new WorkOS('sk_test_Sz3IQjepeSWaI4cMS4ms4sMuU');
        const listConnectionsResponse: ListResponse<ConnectionResponse> = {
          object: 'list',
          data: [connectionResponse],
          list_metadata: {},
        };

        fetchOnce(listConnectionsResponse);

        await workos.sso.listConnections({
          connectionType: ConnectionType.OktaSAML,
          organizationId: 'org_123',
        });

        expect(fetchSearchParams()).toMatchObject({
          connection_type: ConnectionType.OktaSAML,
          organization_id: 'org_123',
        });
      });
    });

    describe('getAuthorizationUrl', () => {
      describe('with no custom api hostname', () => {
        it('generates an authorize url with the default api hostname', () => {
          const workos = new WorkOS('sk_test_Sz3IQjepeSWaI4cMS4ms4sMuU');

          const url = workos.sso.getAuthorizationUrl({
            domain: 'lyft.com',
            clientId: 'proj_123',
            redirectUri: 'example.com/sso/workos/callback',
          });

          expect(url).toMatchSnapshot();
        });
      });

      describe('with no domain or provider', () => {
        it('throws an error for incomplete arguments', () => {
          const workos = new WorkOS('sk_test_Sz3IQjepeSWaI4cMS4ms4sMuU');

          const urlFn = () =>
            workos.sso.getAuthorizationUrl({
              clientId: 'proj_123',
              redirectUri: 'example.com/sso/workos/callback',
            });

          expect(urlFn).toThrowErrorMatchingSnapshot();
        });
      });

      describe('with a provider', () => {
        it('generates an authorize url with the provider', () => {
          const workos = new WorkOS('sk_test_Sz3IQjepeSWaI4cMS4ms4sMuU', {
            apiHostname: 'api.workos.dev',
          });

          const url = workos.sso.getAuthorizationUrl({
            provider: 'Google',
            clientId: 'proj_123',
            redirectUri: 'example.com/sso/workos/callback',
          });

          expect(url).toMatchSnapshot();
        });
      });

      describe('with a connection', () => {
        it('generates an authorize url with the connection', () => {
          const workos = new WorkOS('sk_test_Sz3IQjepeSWaI4cMS4ms4sMuU', {
            apiHostname: 'api.workos.dev',
          });

          const url = workos.sso.getAuthorizationUrl({
            connection: 'connection_123',
            clientId: 'proj_123',
            redirectUri: 'example.com/sso/workos/callback',
          });

          expect(url).toMatchSnapshot();
        });
      });

      describe('with an `organization`', () => {
        it('generates an authorization URL with the organization', () => {
          const workos = new WorkOS('sk_test_Sz3IQjepeSWaI4cMS4ms4sMuU', {
            apiHostname: 'api.workos.dev',
          });

          const url = workos.sso.getAuthorizationUrl({
            organization: 'organization_123',
            clientId: 'proj_123',
            redirectUri: 'example.com/sso/workos/callback',
          });

          expect(url).toMatchSnapshot();
        });
      });

      describe('with a custom api hostname', () => {
        it('generates an authorize url with the custom api hostname', () => {
          const workos = new WorkOS('sk_test_Sz3IQjepeSWaI4cMS4ms4sMuU', {
            apiHostname: 'api.workos.dev',
          });

          const url = workos.sso.getAuthorizationUrl({
            domain: 'lyft.com',
            clientId: 'proj_123',
            redirectUri: 'example.com/sso/workos/callback',
          });

          expect(url).toMatchSnapshot();
        });
      });

      describe('with state', () => {
        it('generates an authorize url with the provided state', () => {
          const workos = new WorkOS('sk_test_Sz3IQjepeSWaI4cMS4ms4sMuU');

          const url = workos.sso.getAuthorizationUrl({
            domain: 'lyft.com',
            clientId: 'proj_123',
            redirectUri: 'example.com/sso/workos/callback',
            state: 'custom state',
          });

          expect(url).toMatchSnapshot();
        });
      });

      describe('with domainHint', () => {
        it('generates an authorize url with the provided domain hint', () => {
          const workos = new WorkOS('sk_test_Sz3IQjepeSWaI4cMS4ms4sMuU');

          const url = workos.sso.getAuthorizationUrl({
            domainHint: 'lyft.com',
            connection: 'connection_123',
            clientId: 'proj_123',
            redirectUri: 'example.com/sso/workos/callback',
            state: 'custom state',
          });

          expect(url).toMatchInlineSnapshot(
            `"https://api.workos.com/sso/authorize?client_id=proj_123&connection=connection_123&domain_hint=lyft.com&redirect_uri=example.com%2Fsso%2Fworkos%2Fcallback&response_type=code&state=custom+state"`,
          );
        });
      });

      describe('with loginHint', () => {
        it('generates an authorize url with the provided login hint', () => {
          const workos = new WorkOS('sk_test_Sz3IQjepeSWaI4cMS4ms4sMuU');

          const url = workos.sso.getAuthorizationUrl({
            loginHint: 'foo@workos.com',
            connection: 'connection_123',
            clientId: 'proj_123',
            redirectUri: 'example.com/sso/workos/callback',
            state: 'custom state',
          });

          expect(url).toMatchInlineSnapshot(
            `"https://api.workos.com/sso/authorize?client_id=proj_123&connection=connection_123&login_hint=foo%40workos.com&redirect_uri=example.com%2Fsso%2Fworkos%2Fcallback&response_type=code&state=custom+state"`,
          );
        });
      });
    });

    describe('getProfileAndToken', () => {
      describe('with all information provided', () => {
        it('sends a request to the WorkOS api for a profile', async () => {
          fetchOnce({
            access_token: '01DMEK0J53CVMC32CK5SE0KZ8Q',
            profile: {
              id: 'prof_123',
              idp_id: '123',
              organization_id: 'org_123',
              connection_id: 'conn_123',
              connection_type: 'OktaSAML',
              email: 'foo@test.com',
              first_name: 'foo',
              last_name: 'bar',
              groups: ['Admins', 'Developers'],
              raw_attributes: {
                email: 'foo@test.com',
                first_name: 'foo',
                last_name: 'bar',
                groups: ['Admins', 'Developers'],
              },
            },
          });

          const workos = new WorkOS('sk_test_Sz3IQjepeSWaI4cMS4ms4sMuU');
          const { accessToken, profile } = await workos.sso.getProfileAndToken({
            code: 'authorization_code',
            clientId: 'proj_123',
          });

          expect(fetch.mock.calls.length).toEqual(1);

          expect(fetchBody()).toMatchSnapshot();
          expect(fetchHeaders()).toMatchSnapshot();
          expect(accessToken).toBe('01DMEK0J53CVMC32CK5SE0KZ8Q');
          expect(profile).toMatchSnapshot();
        });
      });

      describe('without a groups attribute', () => {
        it('sends a request to the WorkOS api for a profile', async () => {
          fetchOnce({
            access_token: '01DMEK0J53CVMC32CK5SE0KZ8Q',
            profile: {
              id: 'prof_123',
              idp_id: '123',
              organization_id: 'org_123',
              connection_id: 'conn_123',
              connection_type: 'OktaSAML',
              email: 'foo@test.com',
              first_name: 'foo',
              last_name: 'bar',
              raw_attributes: {
                email: 'foo@test.com',
                first_name: 'foo',
                last_name: 'bar',
              },
            },
          });

          const workos = new WorkOS('sk_test_Sz3IQjepeSWaI4cMS4ms4sMuU');
          const { accessToken, profile } = await workos.sso.getProfileAndToken({
            code: 'authorization_code',
            clientId: 'proj_123',
          });

          expect(fetch.mock.calls.length).toEqual(1);

          expect(fetchBody()).toMatchSnapshot();
          expect(fetchHeaders()).toMatchSnapshot();
          expect(accessToken).toBe('01DMEK0J53CVMC32CK5SE0KZ8Q');
          expect(profile).toMatchSnapshot();
        });
      });
    });

    describe('getProfile', () => {
      it('calls the `/sso/profile` endpoint with the provided access token', async () => {
        fetchOnce({
          id: 'prof_123',
          idp_id: '123',
          organization_id: 'org_123',
          connection_id: 'conn_123',
          connection_type: 'OktaSAML',
          email: 'foo@test.com',
          first_name: 'foo',
          last_name: 'bar',
          groups: ['Admins', 'Developers'],
          raw_attributes: {
            email: 'foo@test.com',
            first_name: 'foo',
            last_name: 'bar',
            groups: ['Admins', 'Developers'],
          },
        });

        const workos = new WorkOS('sk_test_Sz3IQjepeSWaI4cMS4ms4sMuU');
        const profile = await workos.sso.getProfile({
          accessToken: 'access_token',
        });

        expect(fetch.mock.calls.length).toEqual(1);
        expect(fetchHeaders()).toMatchObject({
          Authorization: 'Bearer access_token',
        });

        expect(profile.id).toBe('prof_123');
      });
    });

    describe('deleteConnection', () => {
      it('sends request to delete a Connection', async () => {
        fetchOnce();

        const workos = new WorkOS('sk_test_Sz3IQjepeSWaI4cMS4ms4sMuU');

        await workos.sso.deleteConnection('conn_123');

        expect(fetchURL()).toContain('/connections/conn_123');
      });
    });

    describe('getConnection', () => {
      it(`requests a Connection`, async () => {
        fetchOnce(connectionResponse);

        const workos = new WorkOS('sk_test_Sz3IQjepeSWaI4cMS4ms4sMuU');

        const subject = await workos.sso.getConnection('conn_123');

        expect(fetchURL()).toContain('/connections/conn_123');

        expect(subject.connectionType).toEqual('OktaSAML');
      });
    });

    describe('listConnections', () => {
      it(`requests a list of Connections`, async () => {
        fetchOnce({
          data: [connectionResponse],
          list_metadata: {},
        });

        const workos = new WorkOS('sk_test_Sz3IQjepeSWaI4cMS4ms4sMuU');

        const subject = await workos.sso.listConnections({
          organizationId: 'org_1234',
        });

        expect(fetchURL()).toContain('/connections');

        expect(subject.data).toHaveLength(1);
      });
    });
  });
});
