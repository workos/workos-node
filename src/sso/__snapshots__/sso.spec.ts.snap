// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SSO SSO getAuthorizationUrl with a connection generates an authorize url with the connection 1`] = `"https://api.workos.dev/sso/authorize?client_id=proj_123&connection=connection_123&redirect_uri=example.com%2Fsso%2Fworkos%2Fcallback&response_type=code"`;

exports[`SSO SSO getAuthorizationUrl with a custom api hostname generates an authorize url with the custom api hostname 1`] = `"https://api.workos.dev/sso/authorize?client_id=proj_123&domain=lyft.com&redirect_uri=example.com%2Fsso%2Fworkos%2Fcallback&response_type=code"`;

exports[`SSO SSO getAuthorizationUrl with a provider generates an authorize url with the provider 1`] = `"https://api.workos.dev/sso/authorize?client_id=proj_123&provider=Google&redirect_uri=example.com%2Fsso%2Fworkos%2Fcallback&response_type=code"`;

exports[`SSO SSO getAuthorizationUrl with an \`organization\` generates an authorization URL with the organization 1`] = `"https://api.workos.dev/sso/authorize?client_id=proj_123&organization=organization_123&redirect_uri=example.com%2Fsso%2Fworkos%2Fcallback&response_type=code"`;

exports[`SSO SSO getAuthorizationUrl with no custom api hostname generates an authorize url with the default api hostname 1`] = `"https://api.workos.com/sso/authorize?client_id=proj_123&domain=lyft.com&redirect_uri=example.com%2Fsso%2Fworkos%2Fcallback&response_type=code"`;

exports[`SSO SSO getAuthorizationUrl with no domain or provider throws an error for incomplete arguments 1`] = `"Incomplete arguments. Need to specify either a 'connection', 'organization', 'domain', or 'provider'."`;

exports[`SSO SSO getAuthorizationUrl with state generates an authorize url with the provided state 1`] = `"https://api.workos.com/sso/authorize?client_id=proj_123&domain=lyft.com&redirect_uri=example.com%2Fsso%2Fworkos%2Fcallback&response_type=code&state=custom+state"`;

exports[`SSO SSO getProfileAndToken with all information provided sends a request to the WorkOS api for a profile 1`] = `"client_id=proj_123&client_secret=sk_test_Sz3IQjepeSWaI4cMS4ms4sMuU&grant_type=authorization_code&code=authorization_code"`;

exports[`SSO SSO getProfileAndToken with all information provided sends a request to the WorkOS api for a profile 2`] = `
{
  "Accept": "application/json, text/plain, */*",
  "Authorization": "Bearer sk_test_Sz3IQjepeSWaI4cMS4ms4sMuU",
  "Content-Type": "application/x-www-form-urlencoded;charset=utf-8",
  "User-Agent": "workos-node/7.6.0",
}
`;

exports[`SSO SSO getProfileAndToken with all information provided sends a request to the WorkOS api for a profile 3`] = `
{
  "connectionId": "conn_123",
  "connectionType": "OktaSAML",
  "email": "foo@test.com",
  "firstName": "foo",
  "groups": [
    "Admins",
    "Developers",
  ],
  "id": "prof_123",
  "idpId": "123",
  "lastName": "bar",
  "organizationId": "org_123",
  "rawAttributes": {
    "email": "foo@test.com",
    "first_name": "foo",
    "groups": [
      "Admins",
      "Developers",
    ],
    "last_name": "bar",
  },
}
`;

exports[`SSO SSO getProfileAndToken without a groups attribute sends a request to the WorkOS api for a profile 1`] = `"client_id=proj_123&client_secret=sk_test_Sz3IQjepeSWaI4cMS4ms4sMuU&grant_type=authorization_code&code=authorization_code"`;

exports[`SSO SSO getProfileAndToken without a groups attribute sends a request to the WorkOS api for a profile 2`] = `
{
  "Accept": "application/json, text/plain, */*",
  "Authorization": "Bearer sk_test_Sz3IQjepeSWaI4cMS4ms4sMuU",
  "Content-Type": "application/x-www-form-urlencoded;charset=utf-8",
  "User-Agent": "workos-node/7.6.0",
}
`;

exports[`SSO SSO getProfileAndToken without a groups attribute sends a request to the WorkOS api for a profile 3`] = `
{
  "connectionId": "conn_123",
  "connectionType": "OktaSAML",
  "email": "foo@test.com",
  "firstName": "foo",
  "groups": undefined,
  "id": "prof_123",
  "idpId": "123",
  "lastName": "bar",
  "organizationId": "org_123",
  "rawAttributes": {
    "email": "foo@test.com",
    "first_name": "foo",
    "last_name": "bar",
  },
}
`;
